### DISCUSSION TITLE
Potential issue with mesh adaptivity and ExtraNodesetGenerator

### DISCUSSION THREAD

**Discussion by starkekr** [idaholab/moose#24369] [Q&A General]
I am having convergence issues with a simple deformation input file (attached).
[m_file_error.txt](https://github.com/idaholab/moose/files/11481965/m_file_error.txt)
In this simulation I have a precipitate which I am translating to one of the corner nodes in the simulation domain (using a time dependent aux variable). The displacement of the  corner nodes are fixed to prevent rigid body translations. After 40 timesteps, the precipitate reaches the corner node and then the simulation stops converging. I am thinking that this is because of mesh adaptivity relabeling the corner node? I also get the same problem with allow_renumbering = false so maybe its something else.

So I am wondering what the default behavior is when an element is refined and that element contains part of a nodeset? Any comments or thought is greatly appreciated :)

**Comment by lindsayad** [comment]
sideset information is associated with the level 0 elements. With respect to nodesets, when new nodes are added along a boundary during mesh refinement, I don't actually know whether they get incorporated into existing nodesets (@roystgnr ?). I would guess not as how could we know the conditions under which a node should be considered as part of a nodeset? If you build a nodeset from a sideset post-adaptivity, then *that* would be clear. Anyway, I don't think that particular information is relevant to you in this case.

What *is* relevant: if you were relying on applying a specific condition to a specific DofObject ID (e.g. element or node ID), then you would indeed be in trouble if renumbering was enabled during adaptivity, but it looks like you are using the nodeset throughout the input file. That should be fine and robust irrespective of whether you allow renumbering or not.

**Reply by roystgnr** [reply]
They don't get incorporated by libMesh.  Nodesets were originally meant to be for things like point sources, pins in elastoplasticity, etc.  "Create a new nodeset for every sideset" was a MOOSE innovation ... so maybe MOOSE updates them after AMR, too?  I'd kinda expect it to be a bug if not, based on how MOOSE codes rely on that.

**Reply by lindsayad** [reply]
Wow you responded before I could even update my post 😄

**Reply by lindsayad** [reply]
Yes we do indeed rely on `build_node_list_from_side_list` post mesh refinement (otherwise our Dirichlet BC classes would not work)

**Reply by starkekr** [reply]
I just realized that I didn't get rid of all of my personal code. The problem still persists if I run this new input file
[m_file_error.txt](https://github.com/idaholab/moose/files/11483035/m_file_error.txt)

**Reply by lindsayad** [reply]
What does poor convergence mean? Poor linear convergence? Poor nonlinear convergence?

**Reply by starkekr** [reply]
Poor nonlinear convergence. Here is a screenshot of the residuals
![image](https://github.com/idaholab/moose/assets/118568957/48e5b5ff-2db1-45c1-bc45-cfba11bdfc4d)
I also just noticed that the displacement field shows problems even earlier than timstep 42. If you compare the x component of the displacement field as shown here in timestep 20
![image](https://github.com/idaholab/moose/assets/118568957/a1a76d36-18b2-4994-a221-57564c406c58)
to the next timestep 21 
![image](https://github.com/idaholab/moose/assets/118568957/2d7bac99-89fb-4388-b828-c9d40107cdfe)
you can see that there are some issues. Both plots are scaled to the same magnitude. I would expect the displacement field to continue to look more or less like the one in timestep 20 but something weird happens as the precipitate approaches the boundary

**Reply by lindsayad** [reply]
Yea I don't know this physics very well. I'm going to ask some of our community members to chime in here

**Reply by lynnmunday** [reply]
will you paste your input file into a comment

**Reply by hugary1995** [reply]
Very likely just bad Jacobian... Are you using AD?

**Reply by starkekr** [reply]
[GlobalParams]
  displacements = 'disp_x disp_y'
[]
[Mesh]
  [msh]
    type = GeneratedMeshGenerator
    dim  = 2
    nx   = 16
    ny   = 16
    nz   = 0
    xmin = 0
    xmax = 4.0
    ymin = 0
    ymax = 4.0
    zmin = 0
    zmax = 0.0
  []
  [ns]
    type         = ExtraNodesetGenerator
    input        = msh
    new_boundary = ns1
    coord        = '0 0 0'
    #nodes        = 1
  []
  uniform_refine    = 3
[]
[Variables]
  [disp_x]
    family = LAGRANGE
    order  = FIRST
  []
  [disp_y]
    family = LAGRANGE
    order  = FIRST
  []
[]

[AuxVariables]
  [eta0]
    family = LAGRANGE
    order  = FIRST
  []
  [eta1]
    family = LAGRANGE
    order  = FIRST
  []
  [bnds]
    family = LAGRANGE
    order  = FIRST
  []
[]
[Kernels]
  [stress_x]
    type              = StressDivergenceTensors
    variable          = disp_x
    component         = 0
    eigenstrain_names = eigenstrain
  []
  [stress_y]
    type              = StressDivergenceTensors
    variable          = disp_y
    component         = 1
    eigenstrain_names = eigenstrain
  []
[]
[AuxKernels]
    [eta0]
		type = FunctionAux
		variable = eta0
		function = '1.0/(1+exp(-(sqrt((x-(1-t))^2+(y-(1-t))^2)-0.5^2)/0.03))
		           +1.0/(1+exp(-(sqrt((x-(5-t))^2+(y-(5-t))^2)-0.5^2)/0.03))
		           +1.0/(1+exp(-(sqrt((x-(1-t))^2+(y-(5-t))^2)-0.5^2)/0.03))
		           +1.0/(1+exp(-(sqrt((x-(5-t))^2+(y-(1-t))^2)-0.5^2)/0.03))
		           -3'
		execute_on = 'initial timestep_begin'
	[]
	[eta1]
		type = ParsedAux
		variable = eta1
		args = eta0
		function = '1-eta0'
		execute_on = 'initial timestep_begin'
	[]
  [bnds]
  	type = ParsedAux
	variable = bnds
	args = 'eta0 eta1'
	function = 'eta0*(1-eta0)'
	execute_on = 'initial timestep_begin'
  []
[]
[Materials]
 [./elasticity_tensor_A]
    type = ComputeElasticityTensor
    base_name = A
    fill_method = symmetric9
    C_ijkl = '1e1 1e0 1e0 1e1 0 1e1 .4e1 .2e1 .5e1'
  [../]
  [./strain_A]
    type = ComputeSmallStrain
    base_name = A
    eigenstrain_names = eigenstrain
  [../]
  [./stress_A]
    type = ComputeLinearElasticStress
    base_name = A
  [../]
  [./eigenstrain_A]
    type = ComputeEigenstrain
    base_name = A
    eigen_base = '0.1 0.05 0 0 0 0.01'
    prefactor = -1
    eigenstrain_name = eigenstrain
  [../]

  [./elasticity_tensor_B]
    type = ComputeElasticityTensor
    base_name = B
    fill_method = symmetric9
    C_ijkl = '1e1 0 0 1e1 0 1e1 .5e1 .5e1 .5e1'
  [../]
  [./strain_B]
    type = ComputeSmallStrain
    base_name = B
    eigenstrain_names = 'B_eigenstrain'
  [../]
  [./stress_B]
    type = ComputeLinearElasticStress
    base_name = B
  [../]
  [./eigenstrain_B]
    type = ComputeEigenstrain
    base_name = B
    eigen_base = '0.1 0.05 0 0 0 0.01'
    prefactor = -1
    eigenstrain_name = 'B_eigenstrain'
  [../]

  [./switching]
    type = SwitchingFunctionMaterial
    eta = eta0
  [../]
  [./combined]
    type = TwoPhaseStressMaterial
    base_A = A
    base_B = B
  [../]
[]
[BCs]
  [fixed_x]
    type     = DirichletBC
    variable = disp_x
    boundary = ns1
    value    = 0
  []
  [fixed_y]
    type     = DirichletBC
    variable = disp_y
    boundary = ns1
    value    = 0
  []
  [Periodic]
    [all]
      variable       = 'disp_x disp_y'
      auto_direction = 'x y'
    []
  []
[]
[Adaptivity]
  marker      = combo
  max_h_level = 4
  [Markers]
    [combo]
      type    = ComboMarker
      markers = interface
    []
    [interface]
      type     = ValueThresholdMarker
      variable = bnds
      refine   = 0.05
      coarsen  = 0.05
    []
  []
[]
[Executioner]
  type                = Transient
  scheme              = bdf2
  solve_type          = NEWTON
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
  petsc_options_value = 'hypre boomeramg 101'
  l_max_its           = 100
  l_tol               = 0.0001
  nl_max_its          = 40
  nl_rel_tol          = 1e-05
  nl_abs_tol          = 1e-07
  dt                  = 0.01
  start_time          = 0.0
  num_steps           = 100
[]
[Outputs]
  exodus                 = true
  execute_on             = 'initial timestep_end'
  print_linear_residuals = false
  checkpoint             = true
  interval               = 1
[]

**Reply by starkekr** [reply]
If you remove the Adaptivity block it runs fine.

**Reply by hugary1995** [reply]
A few suggestions:
1. Use a direct solver. If it works, then I can give you a better hypre solver.
2. Reduce the eigenstrain by a factor of 100.
3. Double check the eigenstrain prefactor.

**Reply by hugary1995** [reply]
4. Use automatic_scaling.

**Reply by maxnezdyur** [reply]
You might also need to decrease your l_tol, it is higher than your nonlinear tolerance.

**Reply by starkekr** [reply]
1. Switching to LU (or ilu) seemed to help it converge in the sense that it now runs to completion, but the displacement field still behaves weird (it grows) as the precipitate approaches the corner node, whereas I think we should just be seeing a a translating field due to the periodicity of the problem. Maybe a better preconditioner would fix this?
 3. I did this. It just rescales the problem by a factor of 100.
 4. I copied this from another input file. I am now using the default value (1.0)

**Reply by hugary1995** [reply]
Displacement periodicity isn't equivalent to strain periodicity, though.

**Comment by starkekr** [comment]
Ahh, yes you are absolutely right, my mistake. With my executioner block set as:
```
[Executioner]
  type                = Transient
  scheme              = bdf2
  solve_type          = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu '
  automatic_scaling   = true
  l_tol               = 1e-12
  nl_max_its          = 40
  nl_rel_tol          = 1e-13
  nl_abs_tol          = 1e-15
  start_time          = 0.0
  dt                  = 0.01
  num_steps           = 150
[]
```
the AMR solution seems to be consistent with the non adaptive case. Just out of curiosity @hugary1995 what preconditioner would you use?

**Reply by hugary1995** [reply]
The hypre boomeramg preconditioner is supposed to be pretty good at solving mechanics problems. However, the default parameters are far less effective compared to a well-tuned set of parameters. This page has a gentle introduction to some of the common parameters: https://mooseframework.inl.gov/releases/moose/v1.0.0/application_development/hypre.html

I've found the following parameters effective for many problems I have at hand:
```python
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_interp_type -pc_hypre_boomeramg_coarsen_type -pc_hypre_boomeramg_agg_nl -pc_hypre_boomeramg_agg_num_paths -pc_hypre_boomeramg_truncfactor'
  petsc_options_value = 'hypre boomeramg 301 0.25 ext+i PMIS 4 2 0.4'
```

I am optimistic that the above preconditioner can get your problem to converge. However, it's NOT going to change the solution. The solution doesn't depend on the type of preconditioner you use, except some numerical corner cases...

**Comment by laagesen** [comment]
I think the problem may be related to pinning the corner node and simultaneously having a misfitting precipitate there - the misfit strain is going to make your displacement want to be non-zero. You are going to have an inconsistency between the physics of the problem and the boundary condition (zero displacement) you are putting on the corner node. Conventional wisdom in the finite element world says you need to pin at least node to prevent finite translations or rotations of the whole mesh, but in previous work on precipitation problems in the MOOSE phase-field module, I have seen that if you use natural (zero-traction) BCs for the mechanical equilibrium equation, you can solve successfully without needing to pin a node. So I would give that a try. If that doesn't work, try pinning a node that is nowhere near a precipitate.

**Reply by hugary1995** [reply]
You are right on the cause of the misfit. Strain periodicity will fix this "problem" without relying on a parallel preconditioner to workaround the null space. We have the entire homogenization system and the global strain system designed for this scenario.

**Reply by starkekr** [reply]
Maybe I'm mistaken but the misfit strain should force the total strain to be nonzero rather than the displacement field. Anyway here is the final input file that seems to be working for those who are interested. It utilizes the global strain system to enforce all the components of the mean stress to 0. @hugary1995  I also tried the preconditioner that you gave and it had trouble converging when the global strain system was added (it worked great without it!) so I reverted back to ilu and it seemed to work for this problem
```
[GlobalParams]
  displacements = 'fluct_x fluct_y'
[]
[Mesh]
  [msh]
    type = GeneratedMeshGenerator
    dim  = 2
    nx   = 16
    ny   = 16
    nz   = 0
    xmin = 0
    xmax = 4.0
    ymin = 0
    ymax = 4.0
    zmin = 0
    zmax = 0.0
  []
  [ns]
    type         = ExtraNodesetGenerator
    input        = msh
    new_boundary = ns1
    coord        = '0 0 0'
    #nodes        = 1
  []
  uniform_refine    = 3
[]
[UserObjects]
  [global_strain_uo]
    type                  = GlobalStrainUserObject
    applied_stress_tensor = '0 0 0 0 0 0'
    execute_on            = 'Initial Linear Nonlinear'
  []
[]
[ScalarKernels]
  [global_strain_kernel]
    type             = GlobalStrain
    variable         = global_strain
    global_strain_uo = global_strain_uo
  []
[]
[Variables]
  [fluct_x]
    family = LAGRANGE
    order  = FIRST
  []
  [fluct_y]
    family = LAGRANGE
    order  = FIRST
  []
  [global_strain]
    order  = THIRD
    family = SCALAR
  []
[]

[AuxVariables]
  [disp_x]
    order   = FIRST
    family  = LAGRANGE
    outputs = exodus
  []
  [disp_y]
    order   = FIRST
    family  = LAGRANGE
    outputs = exodus
  []
  [eta0]
    family = LAGRANGE
    order  = FIRST
  []
  [eta1]
    family = LAGRANGE
    order  = FIRST
  []
  [bnds]
    family = LAGRANGE
    order  = FIRST
  []
[]
[Kernels]
  [stress_x]
    type              = StressDivergenceTensors
    variable          = fluct_x
    component         = 0
    eigenstrain_names = eigenstrain
  []
  [stress_y]
    type              = StressDivergenceTensors
    variable          = fluct_y
    component         = 1
    eigenstrain_names = eigenstrain
  []
[]
[AuxKernels]
  [disp_x]
    type                 = GlobalDisplacementAux
    variable             = disp_x
    scalar_global_strain = global_strain
    global_strain_uo     = global_strain_uo
    component            = 0
  []
  [disp_y]
    type                 = GlobalDisplacementAux
    variable             = disp_y
    scalar_global_strain = global_strain
    global_strain_uo     = global_strain_uo
    component            = 1
  []
    [eta0]
		type = FunctionAux
		variable = eta0
		function = '1.0/(1+exp(-(sqrt((x-(1-t))^2+(y-(1-t))^2)-0.5^2)/0.03))
		           +1.0/(1+exp(-(sqrt((x-(5-t))^2+(y-(5-t))^2)-0.5^2)/0.03))
		           +1.0/(1+exp(-(sqrt((x-(1-t))^2+(y-(5-t))^2)-0.5^2)/0.03))
		           +1.0/(1+exp(-(sqrt((x-(5-t))^2+(y-(1-t))^2)-0.5^2)/0.03))
		           -3'
		execute_on = 'initial timestep_begin'
	[]
	[eta1]
		type = ParsedAux
		variable = eta1
		args = eta0
		function = '1-eta0'
		execute_on = 'initial timestep_begin'
	[]
  [bnds]
  	type = ParsedAux
	variable = bnds
	args = 'eta0 eta1'
	function = 'eta0*(1-eta0)'
	execute_on = 'initial timestep_begin'
  []
[]
[Materials]
  [heta0]
    type              = SwitchingFunctionMultiPhaseMaterial
    h_name            = heta0
    all_etas          = 'eta0 eta1'
    phase_etas        = eta0
    outputs           = exodus
    output_properties = heta0
  []
  [heta1]
    type              = SwitchingFunctionMultiPhaseMaterial
    h_name            = heta1
    all_etas          = 'eta0 eta1'
    phase_etas        = eta1
    outputs           = exodus
    output_properties = heta1
  []
  [C_eta0]
    type            = ComputeElasticityTensor
    fill_method     = symmetric9
    C_ijkl          = '1.025 0.288 0.406 1.025 0.404 1.059 0.183 0.183 0.368'
    base_name       = C_eta0
    rotation_matrix = '-0.98726 0.15910 0.00000 -0.15910 -0.98726 0.00000 0.00000 0.00000 1.00000'
  []
  [e_eta1]
    type            = RotatedGenericConstantRankTwoTensor
    tensor_values   = '-0.138 0.0 0.0 0.0 0.493 0.0 0.0 0.0 -0.064'
    tensor_name     = e_eta1
    rotation_matrix = '-0.54917 -0.83571 0.00000 0.83571 -0.54917 0.00000 -0.00000 0.00000 1.00000'
  []
  [C_eta1]
    type            = ComputeElasticityTensor
    fill_method     = symmetric9
    C_ijkl          = '1.011 0.643 0.680 1.011 0.643 1.036 0.433 0.417 0.417'
    base_name       = C_eta1
    rotation_matrix = '-0.54917 -0.83571 0.00000 0.83571 -0.54917 0.00000 -0.00000 0.00000 1.00000'
  []
  [elastic_tensor]
    type    = CompositeElasticityTensor
    args    = 'eta0 eta1'
    tensors = 'C_eta0 C_eta1'
    weights = 'heta0 heta1'
  []
  [eigenstrain]
    type             = CompositeEigenstrain
    tensors          = 'e_eta1'
    weights          = 'heta1'
    args             = 'eta0 eta1'
    eigenstrain_name = eigenstrain
  []
  [strain]
    type          = ComputeSmallStrain
    eigenstrain_names = eigenstrain
    global_strain = global_strain
  []
  [stress]
    type              = ComputeLinearElasticStress
    outputs           = exodus
    output_properties = stress
  []
  [global_strain_material]
    type                 = ComputeGlobalStrain
    scalar_global_strain = global_strain
    global_strain_uo     = global_strain_uo
  []
[]
[BCs]
  [fixed_x]
    type     = DirichletBC
    variable = fluct_x
    boundary = ns1
    value    = 0
  []
  [fixed_y]
    type     = DirichletBC
    variable = fluct_y
    boundary = ns1
    value    = 0
  []
  [Periodic]
    [all]
      variable       = 'fluct_x fluct_y'
      auto_direction = 'x y'
    []
  []
[]
[Adaptivity]
  marker      = combo
  max_h_level = 4
  [Markers]
    [combo]
      type    = ComboMarker
      markers = interface
    []
    [interface]
      type     = ValueThresholdMarker
      variable = bnds
      refine   = 0.05
      coarsen  = 0.05
    []
  []
[]
[Executioner]
  type                = Transient
  scheme              = bdf2
  solve_type          = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'ilu '
  automatic_scaling   = true
  l_tol               = 1e-12
  nl_max_its          = 40
  nl_rel_tol          = 1e-13
  nl_abs_tol          = 1e-15
  start_time          = 0.0
  dt                  = 0.01
  num_steps           = 150
[]
[Outputs]
  exodus                 = true
  execute_on             = 'initial timestep_end'
  print_linear_residuals = false
  checkpoint             = true
  interval               = 1
[]
```

**Reply by laagesen** [reply]
Thanks for sharing this. You're right that the misfit strain affects the strain field, but strain is calculated as derivatives of displacements, and displacements are actually what are being solved for- you can't get non-zero strain without spatially varying displacements.

**Reply by hugary1995** [reply]
Right, the hypre boomeramg preconditioner doesn't handle saddle point problem well (which you get by using the global strain system). I think a field split approach should work, though I had no luck with it so far. I've heard @ttruster discovered a bug in the FSP and that could fix it potentially.

**Reply by lindsayad** [reply]
There wasn't really a bug. It just wasn't allowed. But yes that is fixed in https://github.com/idaholab/moose/pull/24379
